cmd/scollector/collectors/awsBilling.go

```diff
@@ -0,0 +1,445 @@
+func (b *billingKeyStructure) parseFromObjectKey(key *string, awsBilling *awsBillingConfig) {
+	keyDir, keyFile := filepath.Split(*key)
+	if keyDir == "" || keyFile == "" {
+		return
+	}
+	dirParts := strings.Split(keyDir, "/")
+	if dirParts[0] != awsBilling.bucketPath {
+		return
+	}
+	dateParts := strings.Split(dirParts[2], "-")
+	b.reportName = dirParts[1]
+	b.reportStart, _ = time.Parse(awsBillingReportDateFormat, dateParts[0])
+	b.reportEnd, _ = time.Parse(awsBillingReportDateFormat, dateParts[1])
+	if len(dirParts) == 5 {
+		b.reportID = dirParts[3]
+	}
+	b.fileName = keyFile
+	b.filePath = *key
+}
+
+func stringInSlice(a string, list []string) bool {
+	for _, b := range list {
+		if b == a {
+			return true
+		}
+	}
+	return false
+}
+
+//Process the billing file and spit out a bill that contains the bill header and
+//line items
+func (awsBilling *awsBillingConfig) ReadBillingFile(contents *[]byte) *billBill {
+	//Parse the bill's CSV format
+	r := csv.NewReader(strings.NewReader(string(*contents)))
+	records, err := r.ReadAll()
+	if err != nil {
+		log.Fatal(err)
+	}
+	//This is going to contain our mapping of fields to columns. First line of the
+	//CSV is the field names. If they change in the future (added/deleted/re-org'd)
+	//then we don't need to screw around with static mappings
+	cols := make(map[string]int)
+	//Get the columns from the first row, which are the field names, and throw
+	//them into our struct with their correct position.
+	for colNo, headerVal := range records[0] {
+		cols[strings.ToUpper(headerVal)] = colNo
+	}
+	//Somewhere to store our bill as we parse it
+	var thisBill billBill
+	//Work through the invoice rows from 1 till the end
+	for rowNo, row := range records[1:] {
+		if rowNo == 1 { //This general invoice stuff is the same on every row, so we only need to parse it once
+			thisBill.InvoiceID = val(row, cols, "bill/InvoiceId")
+			thisBill.BillingEntity = val(row, cols, "bill/BillingEntity")
+			thisBill.BillType = val(row, cols, "bill/BillType")
+			thisBill.PayerAccountID = val(row, cols, "bill/PayerAccountId")
+			thisBill.BillingPeriodStartDate, _ = time.Parse(time.RFC3339, val(row, cols, "bill/BillingPeriodStartDate"))
+			thisBill.BillingPeriodEndDate, _ = time.Parse(time.RFC3339, val(row, cols, "bill/BillingPeriodEndDate"))
+		}
+		//Only process lines that have a product code we care about
+		if awsBilling.prodCodesReg.MatchString(val(row, cols, "lineItem/ProductCode")) {
+			//Parse the contents of the row
+			thisBillLine := &billLineItem{}                                //Somewhere to store our row
+			thisBillLine.parseFromBill(row, cols, awsBilling)              //Parse it
+			thisBill.LineItems = append(thisBill.LineItems, *thisBillLine) //Add it to the bill
+		}
+	}
+	return &thisBill
+}
+
+//Return the a given column's value from a row, given a field
+func val(row []string, cols map[string]int, field string) string {
```

Comment:
Checks properly now, but comment didn't collapse due to this specific line not changing